File: src/main/java/dev/px/physics/Util/Math/Vector/Vec2d.java
package dev.px.physics.Util.Math.Vector;

/**
 * @author James
 */
public class Vec2d {

    public double x, y;

    public Vec2d(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public Vec2d add(Vec2d v) {
        this.x += v.x;
        this.y += v.y;
        return this;
    }

    public Vec2d add(double x, double y) {
        this.x += x;
        this.y += y;
        return this;
    }

    public Vec2d subtract(Vec2d v) {
        this.x -= v.x;
        this.y -= v.y;
        return this;
    }

    public Vec2d subtract(double x, double y) {
        this.x -= x;
        this.y -= y;
        return this;
    }

    public Vec2d multiply(double scalar) {
        this.x *= scalar;
        this.y *= scalar;
        return this;
    }

    public Vec2d multiply(Vec2d v) {
        this.x *= v.x;
        this.y *= v.y;
        return this;
    }

    public Vec2d divide(double scalar) {
        this.x /= scalar;
        this.y /= scalar;
        return this;
    }

    public double length() {
        return Math.sqrt(x * x + y * y);
    }

    public double lengthSquared() {
        return x * x + y * y;
    }

    public Vec2d normalize() {
        double len = length();
        if (len != 0) {
            divide(len);
        }
        return this;
    }

    public double dot(Vec2d v) {
        return x * v.x + y * v.y;
    }

    public double distanceTo(Vec2d v) {
        double dx = v.x - x;
        double dy = v.y - y;
        return Math.sqrt(dx * dx + dy * dy);
    }

    public Vec2d copy() {
        return new Vec2d(x, y);
    }

    public Vec2d set(double x, double y) {
        this.x = x;
        this.y = y;
        return this;
    }

    @Override
    public String toString() {
        return "Vec2d{" + "x=" + x + ", y=" + y + '}';
    }

}

File: untitled/src/main/kotlin/Utility/ADT/Vec2d.kt
package design.after.image.Utility.ADT

import kotlin.math.sqrt

/**
 * @author James
 */
data class Vec2d(var x: Float, var y: Float) {
    fun set(x: Float, y: Float): Vec2d {
        this.x = x; this.y = y
        return this
    }

    fun add(other: Vec2d): Vec2d {
        x += other.x; y += other.y
        return this
    }

    fun subtract(other: Vec2d): Vec2d {
        x -= other.x; y -= other.y
        return this
    }

    fun multiply(scalar: Float): Vec2d {
        x *= scalar; y *= scalar
        return this
    }

    fun normalize(): Vec2d {
        val len = sqrt(x * x + y * y)
        if (len != 0f) {
            x /= len; y /= len
        }
        return this
    }
}
